TensorRT profile generation mode

tfeher opened the pull request:
This PR adds a profile generation to dynamic shapes mode. 

This features completes the ground work for dynamic shapes mode. Profile generation mode is enabled during build(), and collects information of the input shapes for TRT engines. This information is used to generate a singe engine with multiple profiles. 

Dynamic shape mode is an experimental feature. The advantage of dynamic shapes is that we can use a single TensorRT engine with multiple profiles to handle input data with different shapes. This saves memory: all input profiles for an engine can share the same weights. Input dimensions that do not influence the number of model parameters can be dynamic (batch size, sequence length, even image dimensions for fully convolutional networks). Currently we generate a separate profile for each input shape that was provided during build mode.

This feature does not support defining profiles during runtime, only during build mode. This makes it important for users to feed data to build() with all interesting shapes so that the single engine can execute those shapes. Currently this feature does not support INT8 calibration.

This feature is not supposed to change any of the existing behavior of TF-TRT in implicit batch mode. Explicit batch mode is changed: it is now required to use build() to generate the engines.

It is expected to use this feature for elementwise ops. We are soon going to enable all the existing converters to work with this feature. An example on how to use this feature:

```python
    root = _model()
    save.save(root, input_saved_model_dir,
              {signature_constants.DEFAULT_SERVING_SIGNATURE_DEF_KEY: root.run})

    rewriter_config_with_trt = rewriter_config_pb2.RewriterConfig()
    rewriter_config_with_trt.optimizers.extend(
        ["constfold", "layout", "constfold"])
    rewriter_config_with_trt.meta_optimizer_iterations = (
        rewriter_config_pb2.RewriterConfig.ONE)
    optimizer = rewriter_config_with_trt.custom_optimizers.add()
    rewriter_config_with_trt.custom_optimizers.add().name = "constfold"
    optimizer.name = "TensorRTOptimizer"
    optimizer.parameter_map["minimum_segment_size"].i = 1
    optimizer.parameter_map["is_dynamic_op"].b = True
    optimizer.parameter_map["maximum_cached_engines"].i = 1
    optimizer.parameter_map["use_implicit_batch"].b = False

    conversion_params = trt_convert.DEFAULT_TRT_CONVERSION_PARAMS._replace(
        rewriter_config_template=rewriter_config_with_trt)
    converter = trt_convert.TrtGraphConverterV2(input_saved_model_dir=input_saved_model_dir,
                                    conversion_params=conversion_params)
    converter.convert()

    input_shapes = [(2, 4), (3, 9), (64, 128)]
    def my_input_fn():
      for x in input_shapes:
        yield (np.random.normal(size=x).astype(np.float32),)

    converter.build(input_fn=my_input_fn)

    output_saved_model_dir = self.mkdtemp()
    converter.save(output_saved_model_dir=output_saved_model_dir)
```

tfeher committed:
Add TRT profile generation mode

tfeher committed:
Run profile generation mode if needed

gbaned commented:
@tfeher Could you please resolve the conflicts? Thanks!

tfeher committed:
Add extra inputs and outputs for BuildParamsWithMask

tfeher committed:
Merge branch 'trt_engine_context_management' into trt_profile_generation_mode

tfeher committed:
Fix bad_function_call

tfeher committed:
Merge branch 'trt_engine_context_management' into trt_profile_generation_mode

tfeher committed:
Fix errors due to function renaming

tfeher committed:
Test optimization profiles in dynamic shape mode

gbaned commented:
@tfeher Still, conflicts are appearing. Can you please resolve? Thanks!

tfeher committed:
Merge branch 'trt_engine_context_management' into trt_profile_generation_mode

tfeher committed:
Merge branch 'trt_engine_context_management' into trt_profile_generation_mode

tfeher committed:
Merge branch 'master' into trt_profile_generation_mode

tfeher committed:
Remove duplicate inpclude line

bixia1 commented on tensorflow/python/compiler/tensorrt/trt_convert.py:
```diff
@@ -1133,8 +1143,50 @@ def build(self, input_fn):
+          [tensor.name for tensor in self._converted_func.outputs])
+      reset_converted_func.graph.structured_outputs = nest.pack_sequence_as(
+          self._converted_func.graph.structured_outputs,
+          reset_converted_func.graph.structured_outputs)
```
This code looks the same as the end of "convert" function to me. Can we make this function return reset_converted_func instead so that we can reuse it for "convert"? Need to rename the variable in this case.

bixia1 commented on tensorflow/python/compiler/tensorrt/trt_convert.py:
```diff
@@ -1133,8 +1143,50 @@ def build(self, input_fn):
+    # after generating all the profiles. The first input is used but any of
+    # the inputs can be used because the shape of this input does not
+    # determine the engine and instead the shapes collected in profiles
+    # determine the engine.
```
This comment is for the if self._need_trt_profiles: block below. Shall we move it closer to there?

bixia1 commented on tensorflow/python/compiler/tensorrt/trt_convert.py:
```diff
@@ -1133,8 +1143,50 @@ def build(self, input_fn):
     for inp in input_fn():
+      if not first_input:
+        first_input = inp
       self._converted_func(*map(ops.convert_to_tensor, inp))
```
I don't understand why we need to run the for loop with _converted_func when if self._need_trt_profiles == true. Can you explain?

tfeher committed:
Use rebuild_func() for conversion too, edit doc

tfeher committed:
Introduce IsExplicitBatchModeEnabled function in trt_convert

tfeher committed:
Disable explicit batch test in trt_engine_op_test

Currently we do not have a build() equivalent in the C++ API, therfore we cannot generate profiles.
Without profiles we cannot test in explicit batch mode (which includes dynamic shapes mode too).

tfeher reviewed (commented):
Hi @bixia1, thank you for the constructive comments! I have updated the PR accordingly.

I had to disable those trt_engine_op tests that use explicit_batch_mode, because build mode (to generate profiles) is not available using the cpp API. We have python tests in trt_mode_test.py to cover these cases.
